Add build_id support for BPF stackmap

vijunag opened the pull request:
Fix #2098.

A separate build_id stackmap can be created with the help of
new macro BPF_STACK_TRACE_BUILDID.
The kernel BPF reports stacktrace in the structure bpf_stack_build_id.
Changes have been made to BPF modules to support resolving symbols
mentioned in this format.
An example tool is also available in examples/tracing/stack_buildid_example.py

yonghong-song commented:
[buildbot, test this please]

yonghong-song commented on src/python/bcc/__init__.py:
```diff
@@ -1262,6 +1288,17 @@ def kprobe_poll(self, timeout = -1):
+      """
+      try:
+        lib.bcc_buildsymcache_add_module(BPF._bsymcache, modname)
+      except Exception, e:
```
It should be `except Exception as e`. This caused a few builtbot errors.

vijunag replied:
I had fixed this in the latest commit on my forked branch. Let me check if it has been merged properly or not.

yonghong-song commented:
@palmtenor could you also take a look at the change?

yonghong-song commented on src/cc/bcc_syms.cc:
```diff
@@ -421,6 +508,45 @@ void bcc_symcache_refresh(void *resolver) {
+  int idx = 0;
+
+  /*cannot resolve in case of fallback*/
+  if (trace->status == BCC_STACK_BUILD_ID_EMPTY or
```
The above `or` probably works. But bcc tends to usee `||` instead. Let us keep it consistent.

yonghong-song commented on src/cc/bcc_syms.cc:
```diff
@@ -421,6 +508,45 @@ void bcc_symcache_refresh(void *resolver) {
+    build_id += "0123456789abcdef"[nib1];
+    build_id += "0123456789abcdef"[nib2];
+    idx++;
+  }
```
Is the above loop for endianness? If this case, it probably won't work for bigendian.

vijunag commented:
Since I do not have access to yhs_dev2, how to address these comments ?

yonghong-song commented:
@vijunag I suggest you take the current patch, rebase on top of the trunk (it should not have conflicts), address the comments, and submit to a new pull request using the same method you did earlier. This way, you will be able to handle any later change request from your end as well.

vijunag commented on src/cc/bcc_syms.cc:
```diff
@@ -378,6 +378,93 @@ bool ProcSyms::Module::find_addr(uint64_t offset, struct bcc_symbol *sym) {
+bool BuildSyms::add_module(const std::string module_name)
+{
+  struct stat s;
+  char buildid[128]; //160 bits seems to be default
```
Saw this code somewhere in bcc_elf_xxx.c. Yes, we should use BCC_BUILD_ID_SIZE.

palmtenor commented:
@yonghong-song Absolutely. Just got home from holiday trip and back having internet:) Will take a look shortly.

vijunag commented:
@yonghong-song - Somehow I cannot squash in my workspace excluding the merge commit. I have addressed the review comments in the forked branch. Please see if you can integrate that here.
